cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,591 @@
+                rshape.append(var.rshape[it])
+                it += 1
+            elif index is None:
+                ashape.append(1)
+                rshape.append(1)
+            elif index is Ellipsis:
+                ashape.extend(var.ashape[it:it + skip])
+                rshape.extend(var.rshape[it:it + skip])
+                it += skip
+
+        ashape.extend(var.ashape[it:var.ndim])
+        rshape.extend(var.rshape[it:var.ndim])
+
+        return self.make_view(
+            var, indexed_from=var, index_key=indices,
+            ashape=tuple(ashape), rshape=tuple(rshape))
+
+    @property
+    def all_variables(self):
+        """Returns the list of all variables this class emitted.
+        """
+        return list(self._variables_dict.values())
+
+
+class _FusionHistory(object):
+    """History of operation exectuted in the target function of fusion.
+    """
+
+    def __init__(self, name):
+        self.name = name
+        self.vc = _VariableConductor()
+        self.shape_constraints = _fusion_shape._ShapeConstraints()
+        self.op_list = []
+        self.submodules = {}
+
+    @staticmethod
+    def _to_interface(x):
+        """Returns an _array or a _scalar object which packs the given value.
+        """
+        if x is None:
+            return None
+        assert isinstance(x, _fusion_variable._FusionCudaVarBase)
+        return x.as_interface()
+
+    def _from_arraylike_interface(self, x):
+        """Returns ``_FusionCuda{Array/Scalar}`` object from the input.
+        """
+        if isinstance(x, _fusion_interface._FusionVariableInterfaceBase):
+            return x.content
+        if isinstance(x, _accepted_types):
+            dtype = numpy.dtype(type(x))
+            return self.vc.generate_new_scalar(dtype, const_value=x)
+        if isinstance(x, (numpy.ndarray, core.ndarray)):
+            raise TypeError('Concrete ndarray is not supported in fusion.')
+        raise TypeError('{} type is not supported'.format(type(x)))
+
+    def _from_interface(self, x):
+        """Returns ``_FusionCuda{Array/Scalar}`` object or ``None``.
+        """
+        if x is None:
+            return None
+        return self._from_arraylike_interface(x)
+
+    def call_ufunc(self, ufunc, *args, **kwargs):
+        """Register an elementwise operation with the given parameters.
+
+        Args:
+            ufunc(_kernel.ufunc): The ufunc to operate.
+            args(tuple): The arguments.
+            kwargs(dict): The keyword arguments.
+        """
+
+        assert isinstance(ufunc, _kernel.ufunc)
+
+        # Parse Inputs.
+        nin = ufunc.nin
+        nout = ufunc.nout
+        dtype = kwargs.pop('dtype', None)
+
+        if 'out' in kwargs and len(args) > nin:
+            raise ValueError(
+                'cannot specify \'out\' as both a positional and '
+                'keyword argument')
+
+        in_params = [self._from_arraylike_interface(x) for x in args[:nin]]
```

Comment:
`in_params` -> `in_args`? because they're aguments rather than parameters.
https://stackoverflow.com/a/156787/402996

Reply:
`in_params` feels better for me because these variables don't have actual array data.

Reply:
How about `in_vars`?
They're not "parameter"s anyway, are they?